# Practice Case Day 1

**MONGO DB USING PYMONGO**

<br>
<br>

In this case, we want to create a new collection based on movies_initial using this mongo client : <br><br>

<b>*mongodb+srv://adminuser12345:adminuser12345@cluster0-phmf4.gcp.mongodb.net/test?retryWrites=true&w=majority*</b>

<br><br>
and change it's format and features to movies using this mongo_client : <br><br>

<b>*mongodb+srv://userstudent:admin1234@cluster0-nnbxe.gcp.mongodb.net/test?retryWrites=true&w=majority*</b>

<br>
We are using pymongo, and use aggregation and named our processed movies_initial : ***clean_movies_andreas***
<br>
*you can check above collection in the first mongo client*

<br>
<br>

In general, our work flow is :
1. Importing the modules
2. Connecting to client
3. Do a little Exploratory collection in Movies, and Movies Initial
4. Do aggregation using pipeline : '$project$ (for changing the field's format) --> $match$ (for query documents) --> $out$ (for create new collection)
5. Validate the result


## Install and Import Module 

<br>
<br>

> We are using pymongo in order to process database in MongoDB using Jupyter Notebook. Pymongo SRV used in order to connect using dnspython. We make sure, that DNSpython already installed by installing dnspython again. 
<br>
We are using !pip install to install those modules


In [1]:
!pip install pymongo
!pip install pymongo[srv]
!pip install dnspython

     |████████████████████████████████| 194kB 2.6MB/s 


<br>
<br>

> After installing the necesary modules, we import those modules. We are using pprint to print dictionaries in a beautiful way. We are using datetime in order to change strings in movies_initial's fields into datetime.<br>

In [0]:

import pymongo
from pymongo import MongoClient
import pprint
from IPython.display import clear_output
import pandas as pd

from datetime import datetime




## Connect to the client

<br><br>
> Because there are no movies in database client, we connect another client

In [0]:
client2 = MongoClient("mongodb+srv://userstudent:admin1234@cluster0-nnbxe.gcp.mongodb.net/test?retryWrites=true&w=majority")
client = MongoClient('mongodb+srv://adminuser12345:adminuser12345@cluster0-phmf4.gcp.mongodb.net/test?retryWrites=true&w=majority')

<br>

> In cluster0, using list_database_names(), we see there are 4 database ; movies_initial, sample_mflix, admin, and local. We use sample_mflix

In [4]:
client.list_database_names()

['movies_initial', 'sample_mflix', 'admin', 'local']

## Exploratory Collection 

<br>

> We assign ***sample_mflix database*** into ***movie_data***. In this database, you can see below each collection in this database

In [5]:
movie_data = client.sample_mflix
movie_data.list_collection_names()

['clean_movies_damar',
 'clean_movies_zumar',
 'clean_movies_andreas',
 'clean_movies_indra_1',
 'clean_movies_frans',
 'clean_movies_firdaus',
 'clean_movies_tunggal',
 'movies_initial',
 'clean_movies_firli']

<br>

> In ***movies_initial***, we see each available fields in all documents. 

In [6]:
clean_movies = movie_data.movies_initial
set([j for i in [doc.keys() for doc in clean_movies.find({})] for j in i])

{'_id',
 'awards',
 'cast',
 'country',
 'director',
 'fullplot',
 'genre',
 'imdbID',
 'imdbRating',
 'imdbVotes',
 'language',
 'lastupdated',
 'metacritic',
 'plot',
 'poster',
 'rating',
 'released',
 'runtime',
 'title',
 'type',
 'writer',
 'year'}

<br><br>

> Below, we see an example of document in ***movies_initial***. We see many empty values in *fields* and *imdb* values still not in the same field.

In [7]:
pprint.pprint(clean_movies.find_one({"imdbID":6517}))

{'_id': ObjectId('5db0085107645f129de5460e'),
 'awards': '1 win.',
 'cast': 'Howard C. Hickman, Enid Markey, Lola May, Kate Bruce',
 'country': 'USA',
 'director': 'Reginald Barker, Thomas H. Ince, Raymond B. West, Walter '
             'Edwards, David Hartford, Jay Hunt, J. Parker Read Jr.',
 'fullplot': 'Allegorical film about peace. A king starts a war, many of the '
             'women are against it, people are pressed into service. A count '
             'has constructed a submarine and gets the order to sink an ocean '
             'liner, that is also carrying - supposedly - ammunition for the '
             'enemy. The count refuses to fire the torpedos, and sinks the '
             'submarine. He survives, but in a limbo between death and life '
             'where he meets Jesus, who takes him over to preach peace. '
             'Naturally the king arrests him and sentences him to death for '
             'treason, but then Jesus shows him the real face of war.',
 'genre': 

<br><br>

> If we see below, we want to change the format of fields in ***movies_initial*** (above) into ***movies*** (below). You can see in *awards, cast,countries,directors*, etc there are a difference between them that is in ***movies***, their values are in a list.<br><br>
> *Imdb* and *Tomatoes* also has a difference. In ***movies Initial*** there are no *viewer, rating etc*
> Also, in *released, runtime*, the format is difference, which is in ***movies*** the format is in datetime while in ***movies_initial***, the format is in string.

In [8]:
sample_mflix = client2.sample_mflix.movies
pprint.pprint(sample_mflix.find_one({"imdb.id":6517}))

{'_id': ObjectId('573a1390f29313caabcd5a93'),
 'awards': {'nominations': 0, 'text': '1 win.', 'wins': 1},
 'cast': ['Howard C. Hickman', 'Enid Markey', 'Lola May', 'Kate Bruce'],
 'countries': ['USA'],
 'directors': ['Reginald Barker',
               'Thomas H. Ince',
               'Raymond B. West',
               'Walter Edwards',
               'David Hartford',
               'Jay Hunt',
               'J. Parker Read Jr.'],
 'fullplot': 'Allegorical film about peace. A king starts a war, many of the '
             'women are against it, people are pressed into service. A count '
             'has constructed a submarine and gets the order to sink an ocean '
             'liner, that is also carrying - supposedly - ammunition for the '
             'enemy. The count refuses to fire the torpedos, and sinks the '
             'submarine. He survives, but in a limbo between death and life '
             'where he meets Jesus, who takes him over to preach peace. '
             'Natura

In [10]:
sample_mflix.find_one({"countries":{"$exists":False}})

{'_id': ObjectId('573a13c6f29313caabd7190a'),
 'awards': {'nominations': 1, 'text': '2 wins & 1 nomination.', 'wins': 2},
 'cast': ['Gillian Anderson',
  'David Arquette',
  'Tillotama Shome',
  'Seema Biswas'],
 'directors': ['Jeffrey D. Brown'],
 'fullplot': 'A girl risks everything for freedom after being trafficked from her mountain village in Nepal to a brothel in India.',
 'genres': ['Drama'],
 'imdb': {'id': 1411956, 'rating': 7.0, 'votes': 112},
 'lastupdated': '2015-08-13 00:36:56.457000000',
 'num_mflix_comments': 1,
 'plot': 'A girl risks everything for freedom after being trafficked from her mountain village in Nepal to a brothel in India.',
 'poster': 'https://m.media-amazon.com/images/M/MV5BMTk3MTQ2NjAwM15BMl5BanBnXkFtZTgwNDY4NjMzMTE@._V1_SY1000_SX677_AL_.jpg',
 'released': datetime.datetime(2014, 3, 7, 0, 0),
 'title': 'Sold',
 'tomatoes': {'lastUpdated': datetime.datetime(2015, 9, 11, 18, 15, 47),
  'viewer': {'numReviews': 15, 'rating': 4.2},
  'website': 'http://www.s

<br>

> First, we must find a way how to query ***movies_initia***l to show only documents that are present in movies. We are using imdb fields, as imdb fields is almost unique. Also, using imdb, we can filter the cast, directors, released, and others, as in imdb there are id (which is unique in all movies), rating, and votes (which is the higher the vote between the same id, the recent the update is). 

<br>

> First we extract imdb fields in all document in movies. As you can see below, the extracted imdb fields are assigned in imdb_list. There are 23541 rows in imdb_list, that is the same length with ***movies*** collection. 

In [11]:
imdb = list(sample_mflix.find({},{'_id':0,'imdb':1}))
imdb_list=[]
for tit in imdb:
  imdb_list.append(list(tit.values())[0])


print(len(imdb_list))

imdb_list[:5]

23541


[{'id': 6517, 'rating': 6.3, 'votes': 162},
 {'id': 8133, 'rating': 7.8, 'votes': 4680},
 {'id': 12512, 'rating': 6.8, 'votes': 489},
 {'id': 13025, 'rating': 7.8, 'votes': 3738},
 {'id': 15498, 'rating': 7.1, 'votes': 327}]

## Aggregation

<br><br>

> Now, we are ready to aggregate. Using below pipeline, we first project ***movies_initial*** fields so that it has the same format using "***movies***". <br>Then, we use match to just pick documents in ***movies initial***, with imdb in movies. <br>Then, use out to create new collection based on this aggregation



<br>
We describe each stage below for further description

In [33]:
client.sample_mflix.drop_collection("clean_movies_andreas")


pipeline = [
        
        

        {
        # Stage 1 -> Project
        '$project': {
            
            # $cond use here so that if there are no title in movies_initial, then we use $nonexistinField that will not showing title 
            'title': {"$cond":[{"$eq":["$title",""]},"$nonExistinField","$title"]},
            'year': {"$cond":[{"$eq":["$year",""]},"$nonExistinField","$year"]},
            'plot': {"$cond":[{"$eq":["$plot",""]},"$nonExistinField","$plot"]},
            
            # Directors,cast etc is a list in movies, so we use split, that automatically split string by commas in our case
            # Then, automaticaly, the output will be on list
            
            'directors': {"$cond":[{"$eq":["$director",""]},"$nonExistinField",{'$split': ["$director", ", "]}]},
            'cast': {"$cond":[{"$eq":["$cast",""]},"$nonExistinField",{'$split': ["$cast", ", "]}]},
            'writers': {"$cond":[{"$eq":["$writer",""]},"$nonExistinField",{'$split': ["$writer", ", "]}]},
            'genres': {"$cond":[{"$eq":["$genre",""]},"$nonExistinField",{'$split': ["$genre", ", "]}]},
            'countries': {"$cond":[{"$eq":["$country",""]},"$nonExistinField",{'$split': ["$country", ", "]}]},
            
            'fullplot': {"$cond":[{"$eq":["$fullplot",""]},"$nonExistinField","$fullplot"]},
            
            # In released, the format is in datetime. We already tried using datetime.strptime but it seems $released is not readable for input in strptime
            'released': {"$cond":[{"$eq":["$released",""]},"$nonExistinField",{ '$concat': [ "datetime.datetime(", 
                                      {'$arrayElemAt':[{'$split': [ "$released", "-" ]}, 0]},
                                      ", ",
                                      {'$arrayElemAt':[{'$split': [ "$released", "-" ]}, 1]},
                                      ", ",
                                      {'$arrayElemAt':[{'$split': [ "$released", "-" ]}, 2]},
                                      ", 0, 0)" ] }]},
            # In movies, there are no string "min" in runtime. So we remove it
            # Condition used so that if there are empty string in documents, it assign into zero
            'runtime': {"$cond": [{'$eq': ["$runtime", ""]}, "$nonExistinField" ,{'$arrayElemAt':[{'$split': [ "$runtime", " min" ]}, 0]}]},
            
            # Metacritic remove metacritic if values in movies_initial "", otherwise show the values
            'metacritic':{"$cond":[{"$eq":["$metacritic",""]},"$nonExistinField","$metacritic"]},
            
            # The values in num_mflix_comments at movies_initial are all "" so we assign it to num_mflix_comments
            'num_mflix_comments': "",
            
            
            'poster': {"$cond":[{"$eq":["$poster",""]},"$nonExistinField","$poster"]},
            'rated': {"$cond":[{"$eq":["$rating",""]},"$nonExistinField","$rating"]},
            
            # imdb values in a movies_initial are grouped into one dictionaries in fields imdb
            'imdb': {
                'id': "$imdbID",
                'rating': "$imdbRating",
                'votes': "$imdbVotes"
                },
            
            # using split and arrayElemAt, we extract strings value in $awards at movies_initial
            'awards': {'nominations': {'$arrayElemAt':[{'$split':[{'$arrayElemAt':[{'$split': [ "$awards", " nomination" ]}, -2]},' ']},-1]},
                       'text': "$awards",
                       'wins': {'$arrayElemAt':[{'$split':[{'$arrayElemAt':[{'$split': [ "$awards", " win" ]}, -2]},' ']},-1]}
                      },
            'type': {"$cond":[{"$eq":["$type",""]},"$nonExistinField","$type"]},
            'lastupdated': {"$cond":[{"$eq":["$lastupdated",""]},"$nonExistinField","$lastupdated"]},
            
            # Tomatoes in movies, are in format dictionaries we split $lastupdated to extract hour,minutes,second
            'tomatoes':{
            'lastupdated':{ '$concat': [ "datetime.datetime(", 
                                      {'$arrayElemAt':[{'$split': [ "$lastupdated", "-" ]}, 0]},
                                      ", ",
                                      {'$arrayElemAt':[{'$split': [ "$lastupdated", "-" ]}, 1]},
                                      ", ",
                                      {'$arrayElemAt':[{'$split': [ {'$arrayElemAt':[{'$split': [ "$lastupdated", " " ]}, 0]} ,"-"]}, 2]},
                                      ", ",{'$arrayElemAt':[{'$split': [ {'$arrayElemAt':[{'$split': [ "$lastupdated", " " ]}, 1]} ,":"]}, 0]},
                                      ", ",{'$arrayElemAt':[{'$split': [ {'$arrayElemAt':[{'$split': [ "$lastupdated", " " ]}, 1]} ,":"]}, 1]},
                                      ", ",{"$arrayElemAt":[{"$split":[{'$arrayElemAt':[{'$split': [ {'$arrayElemAt':[{'$split': [ "$lastupdated", " " ]}, 1]} ,":"]}, 2]},"."]},0]},
                                        ")"] },
                'viewer': {'meter':'', 'numReviews':'', 'rating':''},
                'dvd':'',
                'fresh':'',
                'rotten':'',
                'production':''},
            'languages': {'$split': ["$language", ", "]}
               
    },
    },{
         "$match":{"imdb":{"$in":[dict(sorted(imdb_list[x].items(), key=lambda x: x[0])) for x in range(len(imdb_list))]}}  
       },
    
    
    {
        ### naming the new collection as output ### 
        '$out': "clean_movies_andreas"
    }
]
clean_movies.aggregate(pipeline)

<br>
Here is the explanation :

<br>

1. In some fields, there are some cases with \$cond and \$nonExistinFields. We use this query operator to show the fields when there are values in it, and to hide the fields (using \$nonExistinFields). 
2. In released, tomatoes(last updated) etc, we are manualy using concat in strings, because datetime.strptime("\$released","\%Y-\%m-\%d") somehow failed because \$released is treated as strings in those function. So we use manual instead
3. Match is used after $project because we want only document that has the same imdb in ***movies***
4. We store above pipeline in new collection

<br>
You can see in our data base that clean_movies_andreas is there

In [34]:
movie_data.list_collection_names()

['clean_movies_damar',
 'clean_movies_zumar',
 'clean_movies_andreas',
 'clean_movies_febi',
 'clean_movies_indra_1',
 'clean_movies_frans',
 'clean_movies_firdaus',
 'clean_movies_tunggal',
 'movies_initial',
 'clean_movies_firli']

## Validation

<br>
To validate our result, we want to test if :

1. All fields in ***movies*** and ***clean_movies_andreas*** are the same 
2. The length of ***clean_movies_andreas*** is the same with ***movies***
3. All the Document in ***clean_movies_andreas*** and ***movies*** are the same and has the same order
4. All values are the same in both collections (BUT as ***movies_initia***l has difference values than ***movies***, we do not test this step)

### 1. All field in both movies are the same

> Below code is the fields in movies. There are total 22 fields. 

In [9]:
set([j for i in [doc.keys() for doc in sample_mflix.find({})] for j in i])

{'_id',
 'awards',
 'cast',
 'countries',
 'directors',
 'fullplot',
 'genres',
 'imdb',
 'languages',
 'lastupdated',
 'metacritic',
 'num_mflix_comments',
 'plot',
 'poster',
 'rated',
 'released',
 'runtime',
 'title',
 'tomatoes',
 'type',
 'writers',
 'year'}

<br> 

> And below code is the fields for ***clean_movies_andreas***. You can see that comparing to above results, they are both the same with same amount of fields, 22. From here we see that both collection have equal fields.

In [35]:
set([j for i in [doc.keys() for doc in movie_data.clean_movies_andreas.find({})] for j in i])

{'_id',
 'awards',
 'cast',
 'countries',
 'directors',
 'fullplot',
 'genres',
 'imdb',
 'languages',
 'lastupdated',
 'metacritic',
 'num_mflix_comments',
 'plot',
 'poster',
 'rated',
 'released',
 'runtime',
 'title',
 'tomatoes',
 'type',
 'writers',
 'year'}

<br>

> If we see three example of ***movies*** and ***clean_movies_andreas***, we see that Their field's format is the same. There are some values, in ***clean_movies_andreas*** that has "" values, as if we drop those fields they wil be not included in ***clean_movies_andreas*** as all those fields value in ***clean_movies_andreas*** are " ". <br>Also noted that ***clean_movies_andreas***'s documents ordering is still not the same with movies.

In [14]:
pprint.pprint(list(movie_data.clean_movies_andreas.find({}).limit(3)))

[{'_id': ObjectId('5db0085107645f129de5458b'),
  'awards': {'nominations': None, 'text': '1 win.', 'wins': '1'},
  'cast': ['Charles Kayser', 'John Ott'],
  'countries': ['USA'],
  'directors': ['William K.L. Dickson'],
  'fullplot': 'A stationary camera looks at a large anvil with a blacksmith '
              'behind it and one on either side. The smith in the middle draws '
              'a heated metal rod from the fire, places it on the anvil, and '
              'all three begin a rhythmic hammering. After several blows, the '
              'metal goes back in the fire. One smith pulls out a bottle of '
              'beer, and they each take a swig. Then, out comes the glowing '
              'metal and the hammering resumes.',
  'genres': ['Short'],
  'imdb': {'id': 5, 'rating': 6.2, 'votes': 1189},
  'languages': [''],
  'lastupdated': '2015-08-26 00:03:50.133000000',
  'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
  'rated': 'UNRATED',
  'released':

In [15]:
pprint.pprint(list(sample_mflix.find({}).limit(3)))

[{'_id': ObjectId('573a1390f29313caabcd5a93'),
  'awards': {'nominations': 0, 'text': '1 win.', 'wins': 1},
  'cast': ['Howard C. Hickman', 'Enid Markey', 'Lola May', 'Kate Bruce'],
  'countries': ['USA'],
  'directors': ['Reginald Barker',
                'Thomas H. Ince',
                'Raymond B. West',
                'Walter Edwards',
                'David Hartford',
                'Jay Hunt',
                'J. Parker Read Jr.'],
  'fullplot': 'Allegorical film about peace. A king starts a war, many of the '
              'women are against it, people are pressed into service. A count '
              'has constructed a submarine and gets the order to sink an ocean '
              'liner, that is also carrying - supposedly - ammunition for the '
              'enemy. The count refuses to fire the torpedos, and sinks the '
              'submarine. He survives, but in a limbo between death and life '
              'where he meets Jesus, who takes him over to preach peace. '
  

<br>

### 2. The length of document in both collection is the same

<br><br>
> In below code, you see the difference between length of ***clean_movies_andreas***, and length of ***movies***. In ***clean_movies_andreas*** there are 23539 rows but in ***movies*** there are 23541 rows. There are 2 rows difference between them. We want to see which *imdb* fields that is not the same between both columns  

In [17]:
len(list(movie_data.clean_movies_andreas.find({})))

23539

In [18]:
len(list(client2.sample_mflix.movies.find({})))

23541

<br><br>
> To check which *imdb* fields that are not the same between both collections, we first \$sortByCount each collections by \$imdb to see the count of each \$imdb combinations. \$sortByCount will grouped documents in collection based on certain fields, count the number of certain fields, and sort them by descending. <br><br>grup are aggregate of ***clean_movies_andreas*** and grup2 are aggregated of ***movies***. If there are values in grup2 that are not in grup, then we print the imdb.<br><br>
*noted that we use* \$*project* *in grup2 to sort the order of each imdb dictionaries. The correct order are id, rating, then votes. If we do not use* \$*project as below, the imdb in grup2 will be in rating, votes, id.* 

In [19]:
grup = list(movie_data.clean_movies_andreas.aggregate([{"$sortByCount":"$imdb"}]))
grup2 = list(client2.sample_mflix.movies.aggregate([{"$project":
                                                    {"imdb":{"id":"$imdb.id","rating":"$imdb.rating","votes":"$imdb.votes"}}
                                                    },{"$sortByCount":"$imdb"}]))

for i in grup2:
  if i not in grup:
    print(i)

{'_id': {'id': 5, 'rating': 6.2, 'votes': 1189}, 'count': 3}


<br><br>
Looks like *imdb id* : 5 are the problem here. if we print each document with imdb 5 in both collections (***clean_movies_andreas***, ***movies***) we get :

In [20]:
pprint.pprint(list(client2.sample_mflix.movies.find({"imdb.id":{"$in":[5]}})))


print('\n----------------------------------\n')


pprint.pprint(list(movie_data.clean_movies_andreas.find({"imdb.id":{"$in":[5]}})))

[{'_id': ObjectId('573a1390f29313caabcd4135'),
  'awards': {'nominations': 0, 'text': '1 win.', 'wins': 1},
  'cast': ['Charles Kayser', 'John Ott'],
  'countries': ['USA'],
  'directors': ['William K.L. Dickson'],
  'fullplot': 'A stationary camera looks at a large anvil with a blacksmith '
              'behind it and one on either side. The smith in the middle draws '
              'a heated metal rod from the fire, places it on the anvil, and '
              'all three begin a rhythmic hammering. After several blows, the '
              'metal goes back in the fire. One smith pulls out a bottle of '
              'beer, and they each take a swig. Then, out comes the glowing '
              'metal and the hammering resumes.',
  'genres': ['Short'],
  'imdb': {'id': 5, 'rating': 6.2, 'votes': 1189},
  'lastupdated': '2015-08-26 00:03:50.133000000',
  'num_mflix_comments': 1,
  'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
  'rated': 'UNRATED',
  'released'

<br><br>
> Now we see why there are 2 rows difference between both collection. In ***clean_movies_andreas***, there are only one document with *imdb.id*:5 while in ***movies*** collection there are 3. <br>To make sure that our \$*match* in pipeline above is not wrong, we want to see if in movies_initial, there are more than one document with *imdb.id* : 5. If there are more than one, then our *match* in pipeline is still incorrect

In [21]:
list(movie_data.movies_initial.find({"imdbID":5}))

[{'_id': ObjectId('5db0085107645f129de5458b'),
  'awards': '1 win.',
  'cast': 'Charles Kayser, John Ott',
  'country': 'USA',
  'director': 'William K.L. Dickson',
  'fullplot': 'A stationary camera looks at a large anvil with a blacksmith behind it and one on either side. The smith in the middle draws a heated metal rod from the fire, places it on the anvil, and all three begin a rhythmic hammering. After several blows, the metal goes back in the fire. One smith pulls out a bottle of beer, and they each take a swig. Then, out comes the glowing metal and the hammering resumes.',
  'genre': 'Short',
  'imdbID': 5,
  'imdbRating': 6.2,
  'imdbVotes': 1189,
  'language': '',
  'lastupdated': '2015-08-26 00:03:50.133000000',
  'metacritic': '',
  'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
  'poster': '',
  'rating': 'UNRATED',
  'released': '1893-05-09',
  'runtime': '1 min',
  'title': 'Blacksmith Scene',
  'type': 'movie',
  'writer': '',
  'year': 1893}]

<br><br>
> See, there are only one document. It means that it's not our pipeline \$*match* fault but rather in movies_initial, there are only one documents with *imdb.id*:5. This means that ***movies_initial*** collection is difference not just by number, but by documents as well (which is there are documents in ***movies*** that are not in ***movies_initial***).<br>To fix this, we can insert 2 more documents to ***clean_movies_andreas*** but as we see it will change the instructions that were given to us, we do not fix this problem and move towards instead.<br><br>*What can we conclude is that our collections will has the same length if movies_initial has all documents in movies.*


### 3. All the Document in clean_movies_andreas and movies are the same and has the same order

<br>

> Now, we want to see if all document in ***clean_movies_andreas*** are the same with ***movies***. Same document here means that the fields and the values for some fields between both collection is the same. For example *imdb*, *title*, *cast*, etc but not fields with values that in all movies_initial is " ".<br>Noted that, it's not possible to do the aggregation in above pipeline with same order as ***movies*** (or at least, we haven't find how).<br><br> in order to test step 3, first we sort each collections based on imdb id, imdb rating, and imdb votes. As values in imdb is uniques in each movies (at least there are no same imdb values with different title, cast, etc) when we sort both collection and compare it WITH of course excluding imdb.id : 5, if all documents in the same index are the same, then we can conclude that all document in both collections are the same given the sorted documents.
<br><br>
*notes that we use aggregate rather than sort() as using sort() will exceds the sort's memory limit that is 33mb. To change this limit, we has to use mongodb console(?). We use match to exclude imdb.id :5 and use sort query operators. in below code* : 
- a is sorted collection for ***clean_movies_andreas***
- b is sorted collection for ***movies***

In [0]:
a=list(movie_data.clean_movies_andreas.aggregate([{"$match":{"imdb.id":{"$ne":5}}},{"$sort":{"imdb.id":1,"imdb.rating":1,"imdb.votes":1}}]))


In [57]:
print(len(a))

23538


<br>There are 23538 rows in a sorted collection ***clean_movies_andreas***. Now if we sorted movies using the same technique

In [0]:
b=list(sample_mflix.aggregate([{"$match":{"imdb.id":{"$ne":5}}},{"$sort":{"imdb.id":1,"imdb.rating":1,"imdb.votes":1}}]))

In [59]:
len(b)

23538

<br><br>
> The length between both sorted collection is the same. Now we compare each index (as both sorted collection is in list type) in both sorted collections list. In below code, if there is just one index with difference imdb values between a and b, then the sorted collection has difference documents. If the inverse then the sorted collection has the same documents 

In [68]:
for i in range(len(a)):
    if a[i]['imdb']==b[i]['imdb']:
        if i==(len(a)-1):
          print(str((i+1)*100/(i+1))+" %")   # as they started from 0 so the final i has to add by one 
          print('The sorted collection has the same documents (based on imdb)')
    else:
        pprint.pprint(a[i-1])
        print("\n")
        pprint.pprint(a[i])
        print("\n")
        pprint.pprint(b[i-1])
        print("\n")
        pprint.pprint(b[i])
        print("\n")
        print('The sorted collection has the DIFFERENCE documents (based on imdb)')
        break
    

100.0 %
The sorted collection has the same documents (based on imdb)


<br>

Voila! Our sorted collection has the same documents. If we see the top 5 between both sorted collections below *(ignore \[none none none list])*, we see that both collection has the same documents. It strengthen our evidence that ***movies*** and ***clean_movies_andreas*** has the same documents

In [72]:
[pprint.pprint(a[i]) for i in range(6)]

{'_id': ObjectId('5db0085107645f129de545a6'),
 'awards': {'nominations': None, 'text': '1 win.', 'wins': '1'},
 'cast': ['A.C. Abadie',
          "Gilbert M. 'Broncho Billy' Anderson",
          'George Barnes',
          'Justus D. Barnes'],
 'countries': ['USA'],
 'directors': ['Edwin S. Porter'],
 'fullplot': 'Among the earliest existing films in American cinema - notable '
             'as the first film that presented a narrative story to tell - it '
             'depicts a group of cowboy outlaws who hold up a train and rob '
             "the passengers. They are then pursued by a Sheriff's posse. "
             'Several scenes have color included - all hand tinted.',
 'genres': ['Short', 'Western'],
 'imdb': {'id': 439, 'rating': 7.4, 'votes': 9847},
 'languages': ['English'],
 'lastupdated': '2015-08-13 00:27:59.177000000',
 'num_mflix_comments': '',
 'plot': 'A group of bandits stage a brazen train hold-up, only to find a '
         'determined posse hot on their heels.',
 'p

[None, None, None, None, None, None]

In [70]:
[pprint.pprint(b[i]) for i in range(6)]

{'_id': ObjectId('573a1390f29313caabcd42e8'),
 'awards': {'nominations': 0, 'text': '1 win.', 'wins': 1},
 'cast': ['A.C. Abadie',
          "Gilbert M. 'Broncho Billy' Anderson",
          'George Barnes',
          'Justus D. Barnes'],
 'countries': ['USA'],
 'directors': ['Edwin S. Porter'],
 'fullplot': 'Among the earliest existing films in American cinema - notable '
             'as the first film that presented a narrative story to tell - it '
             'depicts a group of cowboy outlaws who hold up a train and rob '
             "the passengers. They are then pursued by a Sheriff's posse. "
             'Several scenes have color included - all hand tinted.',
 'genres': ['Short', 'Western'],
 'imdb': {'id': 439, 'rating': 7.4, 'votes': 9847},
 'languages': ['English'],
 'lastupdated': '2015-08-13 00:27:59.177000000',
 'plot': 'A group of bandits stage a brazen train hold-up, only to find a '
         'determined posse hot on their heels.',
 'poster': 'https://m.media-amazon.

[None, None, None, None, None, None]

## Conclussion

1. Using pipeline in above code, we can make almost identical collection to ***movies*** based on ***movies_initial***. The problem is that ***movies_initial*** has difference values (not fields, but values) compared to ***movies*** so that to make identical collection is impossible. It's possible though, using \$*lookup* to "join" both collection but because ***movies*** collection is not in the same database as ***movies_initial*** and we can not do cross join between databse, it's impossible.

2. It's also difficult to make collection identical to the ordering of other collection. Unless the ordering in those other collection is known, that we can just add \$*sort* in above pipeline. But as ***movies*** ordering seems like random, we can not.

3. In the end, we are succesfuly makes an identical collection with the same format and almost the same documents